### PR TITLE
fix bootstrap merge

### LINKED ISSUE #2800: bootstrap merge not working
```console
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: eg
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller
  parametersRef:
    group: gateway.envoyproxy.io
    kind: EnvoyProxy
    name: proxy-config
    namespace: envoy-gateway-system
---
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: eg
spec:
  gatewayClassName: eg
  listeners:
    - name: http
      protocol: HTTP
      port: 80
---
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: proxy-config
  namespace: envoy-gateway-system
spec:
  bootstrap:
    type: Merge
    value: |
      stats_sinks:
        - name: envoy.stat_sinks.metrics_service
          typed_config:
            "@type": type.googleapis.com/envoy.config.metrics.v3.MetricsServiceConfig
            transport_api_version: V3
            grpc_service:
              envoy_grpc:
                cluster_name: metrics_cluster
      static_resources:
        clusters:
          - name: metrics_cluster
            type: STRICT_DNS
            lb_policy: ROUND_ROBIN
            load_assignment:
              cluster_name: metrics_cluster
              endpoints:
                - lb_endpoints:
                    - endpoint:
                        address:
                          socket_address:
                            address: skywalking-oap.skywalking.svc.cluster.local
                            port_value: 8080
```

Error message:
```console
2024-03-06T13:46:56.265Z	INFO	provider	controller/controller.go:115	Observed a panic in reconciler: interface conversion: interface {} is nil, not []interface {}	{"runner": "provider", "controller": "gatewayapi", "object": {"name":"gateway.envoyproxy.io/gatewayclass-controller"}, "namespace": "", "name": "gateway.envoyproxy.io/gatewayclass-controller", "reconcileID": "2691f6d1-2b43-40a6-8ae3-0967b4cea3b5"}
panic: interface conversion: interface {} is nil, not []interface {} [recovered]
	panic: interface conversion: interface {} is nil, not []interface {}

goroutine 470 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:116 +0x1a4
panic({0x1bbb000?, 0x400094d350?})
	/opt/hostedtoolcache/go/1.21.7/x64/src/runtime/panic.go:914 +0x218
github.com/envoyproxy/gateway/internal/utils/yaml.mergeMaps(0x4000775440, 0x400074c2c0?)
	/home/runner/work/gateway/gateway/internal/utils/yaml/yaml.go:56 +0x428
github.com/envoyproxy/gateway/internal/utils/yaml.MergeYAML({0x400023c000, 0xf34}, {0x400074c2c0, 0x2ad})
	/home/runner/work/gateway/gateway/internal/utils/yaml/yaml.go:31 +0x12c
github.com/envoyproxy/gateway/internal/xds/bootstrap.ApplyBootstrapConfig(0x0?, {0x400023c000?, 0x400006eb40?})
	/home/runner/work/gateway/gateway/internal/xds/bootstrap/util.go:17 +0x6c
github.com/envoyproxy/gateway/api/v1alpha1/validation.validateBootstrap(0x1fe56c6?)
	/home/runner/work/gateway/gateway/api/v1alpha1/validation/envoyproxy_validate.go:150 +0x58
github.com/envoyproxy/gateway/api/v1alpha1/validation.validateEnvoyProxySpec(0x400058e688)
	/home/runner/work/gateway/gateway/api/v1alpha1/validation/envoyproxy_validate.go:56 +0x1a0
github.com/envoyproxy/gateway/api/v1alpha1/validation.ValidateEnvoyProxy(0x2365218?)
	/home/runner/work/gateway/gateway/api/v1alpha1/validation/envoyproxy_validate.go:33 +0x28
github.com/envoyproxy/gateway/internal/provider/kubernetes.(*gatewayAPIReconciler).processParamsRef(0x400060f8c0, {0x2360638, 0x4000c11650}, 0x4000b6db80, 0x40004f4000)
	/home/runner/work/gateway/gateway/internal/provider/kubernetes/controller.go:1200 +0x658
github.com/envoyproxy/gateway/internal/provider/kubernetes.(*gatewayAPIReconciler).Reconcile(0x400060f8c0, {0x2360638, 0x4000c11650}, {{{0x1fe6e3f?, 0x5?}, {0x100?, 0x4000d19cf8?}}})
	/home/runner/work/gateway/gateway/internal/provider/kubernetes/controller.go:348 +0x9a4
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2365218?, {0x2360638?, 0x4000c11650?}, {{{0x0?, 0xb?}, {0x400067ce10?, 0x0?}}})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119 +0x8c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x40008506e0, {0x2360670, 0x4000846b40}, {0x1cd7f20?, 0x4000834f80?})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316 +0x2e8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x40008506e0, {0x2360670, 0x4000846b40})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266 +0x16c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227 +0x74
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 126
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:223 +0x43c
```

### PR DESCRIPTION
**Pull request by zirain**
Fixes #2800

### CODE REVIEW

**Review by zhaohuabing — APPROVED** [review]

File: internal/xds/bootstrap/util.go [hunk 1]
```diff
@@ -6,19 +6,45 @@
...
+	"github.com/envoyproxy/gateway/internal/utils/proto"
+	_ "github.com/envoyproxy/gateway/internal/xds/extensions" // DON'T REMOVE: import of all extensions
 )
 
 // ApplyBootstrapConfig applies the bootstrap config to the default bootstrap config and return the result config.
 func ApplyBootstrapConfig(boostrapConfig *egv1a1.ProxyBootstrap, defaultBootstrap string) (string, error) {
 	bootstrapType := boostrapConfig.Type
 	if bootstrapType != nil && *bootstrapType == egv1a1.BootstrapTypeMerge {
-		mergedBootstrap, err := yamlutils.MergeYAML(defaultBootstrap, boostrapConfig.Value)
+		mergedBootstrap, err := mergeBootstrap(defaultBootstrap, boostrapConfig.Value)
```
Comment (arkodg) [inline internal/xds/bootstrap/util.go:22]: can you highlight the limitation of the existing function ?

File: internal/utils/proto/google_proto.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+import (
+	"fmt"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+type (
```
Comment (arkodg) [inline internal/utils/proto/google_proto.go:19]: can we add the right code attributions
similar to https://github.com/istio/istio/blob/c71999b6bb65355de5195b4d8ce345f7e47ddc11/pkg/proto/merge/merge.go#L49 ?

File: internal/xds/bootstrap/util.go (hunk 1 above)
Comment (arkodg) [reply internal/xds/bootstrap/util.go:22]: the current code handles the unmarshal and validation later in https://github.com/envoyproxy/gateway/blob/ef1934f1461157afc890cddac863418c3e90b4f0/api/v1alpha1/validation/envoyproxy_validate.go#L155

Comment (zirain) [reply internal/xds/bootstrap/util.go:22]: can you take a look at the linked issue?

Comment (arkodg) [reply internal/xds/bootstrap/util.go:22]: sounds like we panicked in `yaml.mergeMaps`, why not fix that ?

Comment (zirain) [reply internal/xds/bootstrap/util.go:22]: IMO, the best choice is let envoy do that using `---config-path`, this's logic from kuma project is more likely same as envoy did.

Also, I don't want to spend endless type handling in `yaml.mergeMaps`, if someone know that better me, I'm ok to close this.

File: api/v1alpha1/validation/envoyproxy_validate.go [hunk 3]
```diff
@@ -8,20 +8,16 @@ package validation
 import (
 	"errors"
 	"fmt"
-	"net/netip"
-	"reflect"
-
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
```
Comment (arkodg) [inline api/v1alpha1/validation/envoyproxy_validate.go:11]: needs to move below ...

File: api/v1alpha1/validation/envoyproxy_validate.go [hunk 4]
```diff
@@ -8,20 +8,16 @@ package validation
...
 	"github.com/google/go-cmp/cmp"
-	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/testing/protocmp"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
-	"sigs.k8s.io/yaml"
-
-	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
-	"github.com/envoyproxy/gateway/internal/xds/bootstrap"
-	_ "github.com/envoyproxy/gateway/internal/xds/extensions" // register the generated types to support protojson unmarshalling
+	"net/netip"
```
Comment (arkodg) [inline api/v1alpha1/validation/envoyproxy_validate.go:20]: needs to move above ....

File: api/v1alpha1/validation/envoyproxy_validate.go [hunk 5]
```diff
@@ -140,42 +138,33 @@ func validateService(spec *egv1a1.EnvoyProxySpec) []error {
...
 
 func validateBootstrap(boostrapConfig *egv1a1.ProxyBootstrap) error {
+	// Validate user bootstrap config
 	defaultBootstrap := &bootstrapv3.Bootstrap{}
 	// TODO: need validate when enable prometheus?
 	defaultBootstrapStr, err := bootstrap.GetRenderedBootstrapConfig(nil)
 	if err != nil {
 		return err
 	}
+	if err := proto.FromYAML([]byte(defaultBootstrapStr), defaultBootstrap); err != nil {
```
Comment (arkodg) [inline api/v1alpha1/validation/envoyproxy_validate.go:148]: nit: is this needed ?

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: api/v1alpha1/validation/envoyproxy_validate.go (hunk 5 above)
Comment (zirain) [reply api/v1alpha1/validation/envoyproxy_validate.go:148]: for L151?

Comment (arkodg) [reply api/v1alpha1/validation/envoyproxy_validate.go:148]: yah

**Review by zhaohuabing — APPROVED** [review]

**Pull request merged**